File: domain/src/main/java/com/github/anrimian/musicplayer/domain/interactors/player/PlayerCoordinatorInteractor.java
package com.github.anrimian.musicplayer.domain.interactors.player;

import com.github.anrimian.musicplayer.domain.models.composition.source.CompositionSource;
import com.github.anrimian.musicplayer.domain.models.composition.source.LibraryCompositionSource;
import com.github.anrimian.musicplayer.domain.models.player.PlayerState;
import com.github.anrimian.musicplayer.domain.models.player.events.PlayerEvent;
import com.github.anrimian.musicplayer.domain.repositories.UiStateRepository;

import java.util.HashMap;

import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.core.Single;
import io.reactivex.rxjava3.subjects.BehaviorSubject;

public class PlayerCoordinatorInteractor {

    private final PlayerInteractor playerInteractor;
    private final UiStateRepository uiStateRepository;

    private final HashMap<PlayerType, CompositionSource> preparedSourcesMap = new HashMap<>();
    private PlayerType activePlayerType = PlayerType.LIBRARY;

    private final BehaviorSubject<PlayerType> activePlayerTypeSubject = BehaviorSubject.createDefault(activePlayerType);

    public PlayerCoordinatorInteractor(PlayerInteractor playerInteractor,
                                       UiStateRepository uiStateRepository) {
        this.playerInteractor = playerInteractor;
        this.uiStateRepository = uiStateRepository;

        initializePlayerType(activePlayerType);
    }

    public void startPlaying(CompositionSource compositionSource, PlayerType playerType) {
        applyPlayerType(playerType);
        playerInteractor.startPlaying(compositionSource);
    }

    public void play(PlayerType playerType, int delay) {
        applyPlayerType(playerType);
        playerInteractor.play(delay);
    }

    public void setInLoadingState(PlayerType playerType) {
        if (playerType == activePlayerType) {
            playerInteractor.setInLoadingState();
        }
    }

    public void updateSource(CompositionSource source, PlayerType playerType) {
        CompositionSource currentSource = preparedSourcesMap.get(playerType);
        if (source.equals(currentSource)) {
            preparedSourcesMap.put(playerType, source);
        }
        if (playerType == activePlayerType) {
            playerInteractor.updateSource(source);
        }
    }

    public void playOrPause(PlayerType playerType) {
        applyPlayerType(playerType);
        playerInteractor.playOrPause();
    }

    public void stop(PlayerType playerType) {
        if (playerType == activePlayerType) {
            playerInteractor.stop();
        }
    }

    public void pause(PlayerType playerType) {
        if (playerType == activePlayerType) {
            playerInteractor.pause();
        }
    }

    public void reset(PlayerType playerType) {
        preparedSourcesMap.remove(playerType);
        if (playerType == activePlayerType) {
            playerInteractor.reset();
        }
    }

    public void fastSeekForward(PlayerType playerType) {
        if (playerType == activePlayerType) {
            playerInteractor.fastSeekForward();
        } //else update and save position
    }

    public void fastSeekBackward(PlayerType playerType) {
        if (playerType == activePlayerType) {
            playerInteractor.fastSeekBackward();
        } //else update and save position
    }

    public void prepareToPlay(CompositionSource compositionSource, PlayerType playerType) {
        preparedSourcesMap.put(playerType, compositionSource);
        if (playerType == activePlayerType) {
            playerInteractor.prepareToPlay(compositionSource);
        }
    }

    public void onSeekStarted(PlayerType playerType) {
        if (activePlayerType == playerType) {
            playerInteractor.onSeekStarted();
        }
    }

    public boolean onSeekFinished(long position, PlayerType playerType) {
        if (activePlayerType == playerType) {
            playerInteractor.onSeekFinished(position);
            return true;
        } else {
            CompositionSource source = preparedSourcesMap.get(playerType);
            if (source != null) {
                applyPositionChange(source, position);
            }
            return false;
        }
    }

    public Single<Long> getActualTrackPosition(PlayerType playerType) {
        return isPlayerTypeActive(playerType)? playerInteractor.getTrackPosition(): Single.just(-1L);
    }

    public void setPlaybackSpeed(float speed, PlayerType playerType) {
        if (activePlayerType == playerType) {
            playerInteractor.setPlaybackSpeed(speed);
        }
    }

    public Observable<PlayerEvent> getPlayerEventsObservable(PlayerType playerType) {
        return playerInteractor.getPlayerEventsObservable()
                .filter(o -> isPlayerTypeActive(playerType));
    }

    public Observable<Long> getTrackPositionObservable(PlayerType playerType) {
        return playerInteractor.getTrackPositionObservable()
                .filter(o -> isPlayerTypeActive(playerType));
    }

    public Observable<PlayerState> getPlayerStateObservable(PlayerType playerType) {
        return playerInteractor.getPlayerStateObservable()
                .map(state -> isPlayerTypeActive(playerType)? state : PlayerState.PAUSE);
    }

    public boolean isPlayerTypeActive(PlayerType playerType) {
        return activePlayerType == playerType;
    }

    public PlayerState getPlayerState(PlayerType playerType) {
        return isPlayerTypeActive(playerType)? playerInteractor.getPlayerState() : PlayerState.PAUSE;
    }

    public Observable<Boolean> getSpeedChangeAvailableObservable() {
        return playerInteractor.getSpeedChangeAvailableObservable();
    }

    public Observable<PlayerType> getActivePlayerTypeObservable() {
        return activePlayerTypeSubject;
    }

    private void applyPlayerType(PlayerType playerType) {
        if (activePlayerType != playerType) {
            playerInteractor.pause();
            CompositionSource source = preparedSourcesMap.get(playerType);
            if (source != null) {
                playerInteractor.prepareToPlay(source);
            }
            CompositionSource oldSource = preparedSourcesMap.get(activePlayerType);
            if (oldSource != null) {
                //noinspection ResultOfMethodCallIgnored
                playerInteractor.getTrackPosition()
                        .subscribe(position -> applyPositionChange(oldSource, position));
            }

            //not only here
            initializePlayerType(playerType);
            activePlayerType = playerType;
            activePlayerTypeSubject.onNext(activePlayerType);
        }
    }

    private void initializePlayerType(PlayerType playerType) {
        switch (playerType) {
            case LIBRARY: {
                playerInteractor.setPlaybackSpeed(uiStateRepository.getCurrentPlaybackSpeed());
                break;
            }
            case EXTERNAL: {
                playerInteractor.setPlaybackSpeed(1f);

            }
        }
    }

    private void applyPositionChange(CompositionSource source, long position) {
        if (source instanceof LibraryCompositionSource) {
            ((LibraryCompositionSource) source).setTrackPosition(position);
        }
    }
}

File: domain/src/main/java/com/github/anrimian/musicplayer/domain/interactors/player/PlayerCoordinatorInteractor.kt
package com.github.anrimian.musicplayer.domain.interactors.player

import com.github.anrimian.musicplayer.domain.models.composition.source.CompositionSource
import com.github.anrimian.musicplayer.domain.models.player.PlayerState
import com.github.anrimian.musicplayer.domain.models.player.events.PlayerEvent
import com.github.anrimian.musicplayer.domain.repositories.UiStateRepository
import com.github.anrimian.musicplayer.domain.utils.functions.Opt
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.core.Single
import io.reactivex.rxjava3.subjects.BehaviorSubject
import java.util.LinkedList

class PlayerCoordinatorInteractor(
    private val playerInteractor: PlayerInteractor,
    private val uiStateRepository: UiStateRepository
) {

    private val preparedSourcesMap = HashMap<PlayerType, SourceInfo>()
    private val cleanupCallbacksMap = HashMap<PlayerType, () -> Unit>()

    @Volatile
    private var activePlayerType: PlayerType? = null
    private val activePlayerTypeSubject = BehaviorSubject.createDefault<Opt<PlayerType>>(Opt())
    private val playerActivationHistory = LinkedList<PlayerType>()
    private val stateLock = Any()

    fun play(playerType: PlayerType, delay: Long = 0L) {
        synchronized(stateLock) {
            applyPlayerType(playerType)
            playerInteractor.play(delay)
        }
    }

    fun playAfterPrepare(playerType: PlayerType) {
        synchronized(stateLock) {
            applyPlayerType(playerType)
            playerInteractor.playAfterPrepare()
        }
    }

    fun updateSource(source: CompositionSource, playerType: PlayerType) {
        synchronized(stateLock) {
            val currentSourceInfo = preparedSourcesMap[playerType]
            if (currentSourceInfo?.source == source) {
                currentSourceInfo.source = source
            }
            if (playerType == activePlayerType) {
                playerInteractor.updateSource(source)
            }
        }
    }

    fun playOrPause(playerType: PlayerType) {
        synchronized(stateLock) {
            applyPlayerType(playerType)
            playerInteractor.playOrPause()
        }
    }

    fun stop(playerType: PlayerType) {
        synchronized(stateLock) {
            if (playerType == activePlayerType) {
                playerInteractor.stop()
            }
        }
    }

    fun pause(playerType: PlayerType) {
        synchronized(stateLock) {
            if (playerType == activePlayerType) {
                playerInteractor.pause()
            }
        }
    }

    fun error(playerType: PlayerType, throwable: Throwable) {
        synchronized(stateLock) {
            if (playerType == activePlayerType) {
                playerInteractor.error(throwable)
            }
        }
    }

    fun reset(playerType: PlayerType, fallbackToPrevious: Boolean = false) {
        synchronized(stateLock) {
            if (activePlayerType == null) {
                playerInteractor.reset()
                return
            }
            preparedSourcesMap.remove(playerType)
            if (activePlayerType == playerType) {
                if (playerActivationHistory.isNotEmpty()) {
                    playerActivationHistory.removeLast()
                }

                cleanupCallbacksMap[playerType]?.invoke()

                val prevPlayerType = playerActivationHistory.lastOrNull()
                if (prevPlayerType == null || !fallbackToPrevious) {
                    activePlayerType = null
                    activePlayerTypeSubject.onNext(Opt())
                    playerInteractor.reset()
                } else {
                    applyPlayerType(prevPlayerType)
                }
            }
        }
    }

    fun fastSeekForward(playerType: PlayerType): Single<Long> {
        synchronized(stateLock) {
            if (playerType == activePlayerType) {
                return playerInteractor.fastSeekForward()
            }
            //do not update position if player is not active
            return getActualTrackPosition(playerType)
        }
    }

    fun fastSeekBackward(playerType: PlayerType): Single<Long> {
        synchronized(stateLock) {
            if (playerType == activePlayerType) {
                return playerInteractor.fastSeekBackward()
            }
            //do not update position if player is not active
            return getActualTrackPosition(playerType)
        }
    }

    fun prepareToPlay(
        compositionSource: CompositionSource,
        playerType: PlayerType,
        startPosition: Long
    ) {
        synchronized(stateLock) {
            preparedSourcesMap[playerType] = SourceInfo(compositionSource, startPosition)
            if (playerType == activePlayerType) {
                playerInteractor.prepareToPlay(compositionSource, startPosition)
            } else if (activePlayerType == null) {
                applyPlayerType(playerType)
            }
        }
    }

    fun onSeekStarted(playerType: PlayerType) {
        synchronized(stateLock) {
            if (activePlayerType == playerType) {
                playerInteractor.onSeekStarted()
            }
        }
    }

    fun onSeekFinished(position: Long, playerType: PlayerType) {
        synchronized(stateLock) {
            if (activePlayerType == playerType) {
                playerInteractor.onSeekFinished(position)
            } else {
                preparedSourcesMap[playerType]?.trackPosition = position
            }
        }
    }

    fun setPlaybackSpeed(speed: Float, playerType: PlayerType) {
        synchronized(stateLock) {
            if (activePlayerType == playerType) {
                playerInteractor.setPlaybackSpeed(speed)
            }
        }
    }

    fun registerCleanupCallback(playerType: PlayerType, callback: () -> Unit) {
        synchronized(stateLock) {
            cleanupCallbacksMap[playerType] = callback
        }
    }

    fun getPlayerEventsObservable(playerType: PlayerType): Observable<PlayerEvent> {
        return playerInteractor.getPlayerEventsObservable()
            .filter { isPlayerTypeActive(playerType) }
    }

    fun getTrackPositionObservable(playerType: PlayerType): Observable<Long> {
        return playerInteractor.getTrackPositionObservable()
            .filter { isPlayerTypeActive(playerType) }
    }

    fun getTrackPositionChangeObservable(playerType: PlayerType): Observable<Long> {
        return playerInteractor.getTrackPositionChangeObservable()
            .filter { isPlayerTypeActive(playerType) }
    }

    fun getPlayerStateObservable(playerType: PlayerType): Observable<PlayerState> {
        return playerInteractor.getPlayerStateObservable()
            .map { state -> if (isPlayerTypeActive(playerType)) state else PlayerState.IDLE }
    }

    fun getIsPlayingStateObservable(playerType: PlayerType): Observable<Boolean> {
        return playerInteractor.getIsPlayingStateObservable()
            .map { state -> if (isPlayerTypeActive(playerType)) state else false }
    }

    fun isPlayerTypeActive(playerType: PlayerType): Boolean {
        return activePlayerType == playerType
    }

    fun getSpeedChangeAvailableObservable() = playerInteractor.getSpeedChangeAvailableObservable()

    fun getActivePlayerTypeObservable(): Observable<PlayerType> {
        return activePlayerTypeSubject.flatMap { opt ->
            val value = opt.value
            return@flatMap if (value != null) {
                Observable.just(value)
            } else {
                Observable.never()
            }
        }
    }

    fun getActualTrackPosition(playerType: PlayerType): Single<Long> {
        synchronized(stateLock) {
            return if (isPlayerTypeActive(playerType)) {
                playerInteractor.getTrackPosition()
            } else {
                Single.just(preparedSourcesMap[playerType]?.trackPosition ?: -1L)
            }
        }
    }

    @Suppress("CheckResult")
    private fun applyPlayerType(playerType: PlayerType) {
        if (activePlayerType == playerType) {
            return
        }
        if (activePlayerType != null) {
            playerInteractor.pause()
            val oldSource = preparedSourcesMap[activePlayerType]
            if (oldSource != null) {
                playerInteractor.getTrackPosition()
                    .subscribe { position -> oldSource.trackPosition = position }
            }
            cleanupCallbacksMap[activePlayerType]?.invoke()
        } else {
            cleanupCallbacksMap.forEach { (type, callback) ->
                if (type != playerType) {
                    callback.invoke()
                }
            }
        }

        initializePlayerType(playerType)
        activePlayerType = playerType
        playerActivationHistory.remove(playerType)
        playerActivationHistory.add(playerType)
        activePlayerTypeSubject.onNext(Opt(activePlayerType))

        val sourceInfo = preparedSourcesMap[playerType]
        if (sourceInfo != null) {
            playerInteractor.prepareToPlay(sourceInfo.source, sourceInfo.trackPosition)
        }
    }

    private fun initializePlayerType(playerType: PlayerType) {
        when (playerType) {
            PlayerType.LIBRARY -> {
                playerInteractor.setPlaybackSpeed(uiStateRepository.currentPlaybackSpeed)
            }
            PlayerType.EXTERNAL -> {
                playerInteractor.setPlaybackSpeed(1f)
            }
        }
    }

    private inner class SourceInfo(
        var source: CompositionSource,
        var trackPosition: Long
    )

}